pkg/kube/controllers/quarksstatefulset/active_passive_reconciler.go

```diff
@@ -0,0 +1,352 @@
+			return err
+		}
+		ctxlog.WithEvent(qSts, "active-passive").Debugf(
+			ctx,
+			"pod %s promoted to active",
+			p.Name,
+		)
+	case pl > 0:
+		for _, p := range labelledPods {
+			ctxlog.WithEvent(qSts, "active-passive").Debugf(
+				ctx,
+				"validating probe in active pod: %s",
+				p.Name,
+			)
+			if probeSucceeded, _ := r.execContainerCmd(&p, container, probeCmd); !probeSucceeded {
+				l := p.GetLabels()
+				delete(l, qstsv1a1.LabelActiveContainer)
+				err = r.client.Update(r.ctx, &p)
+				if err != nil {
+					return err
+				}
+				ctxlog.WithEvent(qSts, "active-passive").Debugf(
+					ctx,
+					"pod %s promoted to passive",
+					p.Name,
+				)
+			}
+		}
+	}
+	return nil
+}
+
+func (r *ReconcileStatefulSetActivePassive) getActivePod(pods *corev1.PodList, container string, cmd []string) (*corev1.Pod, error) {
+	for _, pod := range pods.Items {
+		if succeed, _ := r.execContainerCmd(&pod, container, cmd); succeed && podutil.IsPodReady(&pod) {
+			return &pod, nil
+		}
+	}
+	return nil, errors.Errorf("executing cmd: %v, in container: %s, failed\n", cmd, container)
+}
+
+func (r *ReconcileStatefulSetActivePassive) getActiveLabels(pods *corev1.PodList, container string) []corev1.Pod {
+	var labelledPods []corev1.Pod
+	for _, pod := range pods.Items {
+		labels := pod.GetLabels()
+		if _, found := labels[qstsv1a1.LabelActiveContainer]; found {
+			labelledPods = append(labelledPods, pod)
+		}
+	}
+	return labelledPods
+}
+
+func (r *ReconcileStatefulSetActivePassive) execContainerCmd(pod *corev1.Pod, container string, command []string) (bool, error) {
+	req := r.kclient.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Name(pod.Name).
+		Namespace(pod.Namespace).
+		SubResource("exec").
+		VersionedParams(&corev1.PodExecOptions{
+			Container: container,
+			Command:   command,
+			Stdin:     true,
+			Stdout:    true,
+			Stderr:    false,
+		}, clientscheme.ParameterCodec)
+
+	executor, err := remotecommand.NewSPDYExecutor(r.kubeconfig, "POST", req.URL())
+	if err != nil {
+		return false, errors.New("failed to initialize remote command executor")
+	}
+	if err = executor.Stream(remotecommand.StreamOptions{Stdin: os.Stdin, Stdout: os.Stdout, Tty: false}); err != nil {
+		return false, errors.Wrapf(err, "failed executing command in pod: %s, container: %s in namespace: %s",
+			pod.Name,
+			container,
+			pod.Namespace,
+		)
+	}
+	ctxlog.Info(r.ctx, "Succesfully exec cmd in container: ", container, ", inside pod: ", pod.Name)
+
+	return true, nil
+}
+
+func (r *ReconcileStatefulSetActivePassive) getStsPodList(ctx context.Context, desiredSts *appsv1.StatefulSet) (*corev1.PodList, error) {
+	podList := &corev1.PodList{}
+	err := r.client.List(ctx,
+		podList,
+		crc.InNamespace(desiredSts.Namespace),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return podList, nil
+}
+
+func (r *ReconcileStatefulSetActivePassive) listStatefulSetVersions(ctx context.Context, qStatefulSet *qstsv1a1.QuarksStatefulSet) (map[int]bool, error) {
```

Comment:
`getStatefulSet`

should be a helper of qsts

Reply:
we do not longer need this function, I removed it